File: frontend/src/types/blueprint.ts
export interface ItemCategory {
  uuid: string;
  record_key: string | null;
  label: string;
  sort_order: number | null;
}

export interface BlueprintSummary {
  uuid: string;
  key: string | null;
  category_uuid: string | null;
  category_label: string | null;
  output_name: string | null;
  output_class: string | null;
  craft_time_label: string | null;
  ingredient_count: number | null;
}

export interface BlueprintIngredient {
  name: string;
  kind: string | null;
  resource_type_uuid: string | null;
  item_uuid: string | null;
  quantity_scu: number | null;
  quantity: number | null;
}

export interface BlueprintOwner {
  user_id: number;
  discord_username: string;
  global_name: string | null;
  server_nickname: string | null;
}

export interface BlueprintDetail extends BlueprintSummary {
  craft_time_seconds: number | null;
  owners: BlueprintOwner[];
  ingredients: BlueprintIngredient[] | null;
}

export interface OrgBlueprint extends BlueprintSummary {
  owner_count: number;
  owners: BlueprintOwner[];
}

export interface UserBlueprintEntry extends BlueprintSummary {
  added_at: string;
}

export interface SyncResult {
  success: boolean;
  result: {
    categories:  { added: number; updated: number };
    blueprints:  { added: number; updated: number };
    ingredients: { synced: number };
  };
}

File: api/database/models/blueprint.py
from sqlalchemy import Column, Integer, String, Float, DateTime, ForeignKey, UniqueConstraint
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from ..connection import Base


class ItemCategory(Base):
    __tablename__ = "item_categories"

    uuid        = Column(String(36), primary_key=True)
    record_key  = Column(String(128), nullable=True)
    label       = Column(String(128), nullable=False)
    sort_order  = Column(Integer, nullable=True)

    blueprints  = relationship("Blueprint", back_populates="category")


class Blueprint(Base):
    __tablename__ = "blueprints"

    uuid               = Column(String(36), primary_key=True)
    key                = Column(String(255), nullable=True)
    category_uuid      = Column(String(36), ForeignKey("item_categories.uuid"), nullable=True)
    output_item_uuid   = Column(String(36), nullable=True)
    output_name        = Column(String(255), nullable=True)
    output_class       = Column(String(128), nullable=True)
    craft_time_seconds = Column(Integer, nullable=True)
    craft_time_label   = Column(String(32), nullable=True)
    ingredient_count   = Column(Integer, nullable=True)
    synced_at          = Column(DateTime(timezone=True), server_default=func.current_timestamp())

    category        = relationship("ItemCategory", back_populates="blueprints")
    user_blueprints = relationship("UserBlueprint", back_populates="blueprint")
    ingredients     = relationship("BlueprintIngredient", back_populates="blueprint", cascade="all, delete-orphan")


class BlueprintIngredient(Base):
    __tablename__ = "blueprint_ingredients"

    id                 = Column(Integer, primary_key=True, autoincrement=True)
    blueprint_uuid     = Column(String(36), ForeignKey("blueprints.uuid", ondelete="CASCADE"), nullable=False, index=True)
    name               = Column(String(255), nullable=False)
    kind               = Column(String(32), nullable=True)
    resource_type_uuid = Column(String(36), nullable=True)
    item_uuid          = Column(String(36), nullable=True)
    quantity_scu       = Column(Float, nullable=True)
    quantity           = Column(Integer, nullable=True)

    blueprint = relationship("Blueprint", back_populates="ingredients")


class UserBlueprint(Base):
    __tablename__ = "user_blueprints"
    __table_args__ = (UniqueConstraint("user_id", "blueprint_uuid", name="uq_user_blueprint"),)

    id            = Column(Integer, primary_key=True, autoincrement=True)
    user_id       = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    blueprint_uuid = Column(String(36), ForeignKey("blueprints.uuid", ondelete="CASCADE"), nullable=False)
    created_at    = Column(DateTime(timezone=True), server_default=func.current_timestamp())

    blueprint = relationship("Blueprint", back_populates="user_blueprints")
